openmeter/productcatalog/subscription/service/change_test.go

```diff
@@ -0,0 +1,258 @@
+
+			// Let's create the subscription
+			p1Inp := plansubscription.PlanInput{}
+			p1Inp.FromRef(&plansubscription.PlanRefInput{
+				Key:     plan1.ToCreateSubscriptionPlanInput().Plan.Key,
+				Version: &plan1.ToCreateSubscriptionPlanInput().Plan.Version,
+			})
+
+			sub, err := svc.Create(ctx, plansubscription.CreateSubscriptionRequest{
+				PlanInput: p1Inp,
+				WorkflowInput: subscription.CreateSubscriptionWorkflowInput{
+					ChangeSubscriptionWorkflowInput: subscription.ChangeSubscriptionWorkflowInput{
+						Name:       "test",
+						ActiveFrom: now.Add(time.Second),
+					},
+					Namespace:  cust.Namespace,
+					CustomerID: cust.ID,
+				},
+			})
+			require.Nil(t, err)
+
+			p2Input := examplePlanInput1
+			p2Input.Plan.PlanMeta.Name = "New Plan"
+			p2Input.Plan.PlanMeta.Key = "new_plan"
+			p2Input.Plan.Phases = append(p2Input.Plan.Phases, productcatalog.Phase{
+				PhaseMeta: productcatalog.PhaseMeta{
+					Key:         "test_phase_4",
+					Name:        "Test Phase 4",
+					Description: lo.ToPtr("Test Phase 4 Description"),
+					StartAfter:  testutils.GetISODuration(t, "P8M"),
+				},
+				RateCards: productcatalog.RateCards{
+					&subscriptiontestutils.ExampleRateCard1,
+				},
+			})
+
+			// Let's create a second plan
+			plan2, err := deps.subDeps.PlanService.CreatePlan(ctx, p2Input)
+			require.Nil(t, err)
+
+			eFrom := clock.Now().Add(5 * time.Second)
+
+			// Let's publish the new plan
+			plan2, err = deps.subDeps.PlanService.PublishPlan(ctx, plan.PublishPlanInput{
+				NamespacedID: plan2.NamespacedID,
+				EffectivePeriod: productcatalog.EffectivePeriod{
+					EffectiveFrom: &eFrom,
+				},
+			})
+			require.Nil(t, err)
+			require.NotNil(t, plan2)
+
+			// Let's create a new version of the second plan
+			p2v2Input := p2Input
+			p2v2Input.Plan.PlanMeta.Name = "New Plan 2"
+
+			plan2v2, err := deps.subDeps.PlanService.CreatePlan(ctx, p2v2Input)
+			require.Nil(t, err)
+
+			eFrom2 := clock.Now().Add(10 * time.Second)
+
+			// Let's publish the new version of the second plan
+			_, err = deps.subDeps.PlanService.PublishPlan(ctx, plan.PublishPlanInput{
+				NamespacedID: plan2v2.NamespacedID,
+				EffectivePeriod: productcatalog.EffectivePeriod{
+					EffectiveFrom: &eFrom2,
+				},
+			})
+			require.Nil(t, err)
+
+			clock.SetTime(eFrom2.Add(time.Second))
+
+			// And let's try to change to the old plan still
+			pInp := plansubscription.PlanInput{}
+			pInp.FromRef(&plansubscription.PlanRefInput{
+				Key:     plan2.Key,
+				Version: &plan2.Version,
+			})
+
+			// Let's change the subscription to the new plan
+			_, err = svc.Change(ctx, plansubscription.ChangeSubscriptionRequest{
+				ID: sub.NamespacedID,
+				WorkflowInput: subscription.ChangeSubscriptionWorkflowInput{
+					ActiveFrom: clock.Now(),
+					Name:       sub.Name,
+				},
+				PlanInput: pInp,
+			})
+
+			require.NotNil(t, err)
+			require.ErrorAs(t, err, lo.ToPtr(&models.GenericUserError{}))
```

Comment:
Should we use typed error? Otherwise this will return 500

Reply:
what do you mean? this is a typed error